core/src/avm2/array.rs

```diff
@@ -11,8 +15,197 @@ use std::{cmp::max, ops::RangeBounds};
 /// the prototype.
 #[derive(Clone, Collect)]
 #[collect(no_drop)]
-pub struct ArrayStorage<'gc> {
-    storage: Vec<Option<Value<'gc>>>,
+pub enum ArrayStorage<'gc> {
+    Dense(Vec<Option<Value<'gc>>>, usize),
```

Comment:
We need documentation here to explain why `Dense` now has a separate length parameter. Looking through the rest of the code, it *appears* that this length parameter counts the number of non-holes (akin to `BTreeMap.len()`.)

Reply:
I think I agree. Probably better to change this to a struct like `Dense { storage: ..., dense_used: size }`. Actually the name `dense_used` feels a bit weird as well.

Reply:
+1 on making it structs (assuming we wanna stay with a BTreeMap)
Also, I'm assuming this means that Dense can't have holes on either end (why not?)? If so, can this be documented too?